### PR TITLE
Enhance pipeline with more options for environment variables

### LINKED ISSUE #71: Add more options to define environment variables (global/project/step)
It would be very cool to have an option to define environment variables on a global level and project level in the pipeline file.
Buildkite had that option in version 2 to have global env vars but they removed it in `v3` for some reason.

I have something in mind that could look like this:
```
projects:
  - label: project-xyz
    path: path/to/my/project
    env:
      THIS_IS_A_PROJECT_SCOPED_ENV_VAR: def456

env:
  - THIS_IS_A_PIPELINE_GLOBAL_ENV_VAR: abc123

steps:
  - label: ":docker: build - "
    command: echo "this is a build step"
    env:
      THIS_IS_A_STEP_SCOPED_ENV_VAR: xyz999

```

Buildpipe could already take care of the double dollar signs needed for some of those variables depending on where they are defined in the pipeline.

### PR DESCRIPTION
**Pull request by mowies**
### This PR
- adds the `env:` key in the pipeline definition to set global env vars that are added to every step
- adds the possibility to add project-wide env vars that are added to all steps of a project
- makes the unit testing more reliable (`assert_line` checks if the line is *anywhere* in the output but does not check where exactly or how often)

This PR fixes #71

### CODE REVIEW

**Comment by mowies** [comment]
@ksindi if you have time, please have a quick look at my implementation.
Behaviour-wise I went with adding pipeline-global env vars to steps only if an `env:` is already defined for that step. I am not quite sure if that would make sense though. Maybe adding the global env vars to every step makes more sense, regardless of their specific configuration

File: main.go
Comment (ksindi) [inline main.go:16]: Nit: we should make the third column line up

**Comment by ksindi** [comment]
This looks great @mowies! I'd also like to get feedback from @emptyflash who has approval/merge authority.

**Comment by ksindi** [comment]
https://github.com/jwplayer/buildpipe-buildkite-plugin/issues/71

File: main.go
Comment (mowies) [reply main.go:16]: done ;)

**Review by boyntoni — APPROVED** [review]

**Comment by mowies** [comment]
I'd love to get some more feedback on the thoughts I posted in my last comment :)


> I am not quite sure if that makes sense though. Maybe adding the global env vars to every step makes more sense, regardless of their specific configuration

^^ this ^^

**Comment by gerardo** [comment]
Hi, @mowies. Are you still working on this feature? I'm kinda stuck using buildpipe 0.6.0 because it's the last version that still supports environment variables. For what it's worth, the PR looks good so far 👍

**Comment by mowies** [comment]
Hi! I'd love to finish this and get it merged but I didn't get feedback so far so I don't know...

**Comment by boyntoni** [comment]
@mowies I will defer to @ksindi, but I personally feel that the global env. should be applied to each step regardless of whether that step has addditional steps defined. The step specific env vars should be supplementary, in my opinion.

**Comment by ksindi** [comment]
This feature makes sense.

**Comment by mowies** [comment]
Cool! I will try to finish this on the weekend

**Comment by mowies** [comment]
@ksindi @boyntoni @emptyflash this PR is ready for review now :)

**Comment by mowies** [comment]
(bump)

**Review by boyntoni — APPROVED** [review]
@mowies I'm sorry I didn't see this earlier. Will wait for @kamil's approval and then merge. Thanks a lot for the contribution!

**Comment by ksindi** [comment]
Approved

**Comment by boyntoni** [comment]
@mowies Do you mind updating the changelog and bumping the version?

**Comment by mowies** [comment]
Yes I can do that. I will also try to add some documentation for the new feature

**Comment by mowies** [comment]
@boyntoni Done ✔️

**Comment by boyntoni** [comment]
Thanks again for the contribution @mowies. This has now been released.

**Pull request merged**
